aten/src/ATen/native/AdaptiveMaxPooling2d.cpp

```diff
@@ -0,0 +1,362 @@
+      }
+    }
+  }
+}
+
+Tensor& adaptive_max_pool2d_backward_out_cpu_template(
+          Tensor& gradInput,
+          const Tensor& gradOutput_,
+          const Tensor& input,
+          const Tensor& indices)
+{
+  int dimW = 2;
+  int dimH = 1;
+  int64_t sizeB = 1;
+  int sizeD;
+  int isizeH;
+  int isizeW;
+  int osizeH;
+  int osizeW;
+
+  /* get contiguous gradOutput */
+  auto gradOutput = gradOutput_.contiguous();
+
+  /* resize */
+  gradInput.resize_as_(input);
+  gradInput.zero_();
+
+  if (input.ndimension() == 4) {
+    sizeB = input.size(0);
+    dimW++;
+    dimH++;
+  }
+
+  sizeD  = input.size(dimH-1);
+  isizeH = input.size(dimH);
+  isizeW = input.size(dimW);
+  osizeH = gradOutput.size(dimH);
+  osizeW = gradOutput.size(dimW);
+
+  /* backprop */
+  if (input.ndimension() == 3)
+  {
+    AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d_backward", [&] {
+      /* get raw pointers */
+      scalar_t *gradInput_data = gradInput.data<scalar_t>();
+      scalar_t *gradOutput_data = gradOutput.data<scalar_t>();
+      int64_t *indices_data = indices.data<int64_t>();
+
+      adaptive_max_pool2d_backward_out_frame<scalar_t>(gradInput_data, gradOutput_data,
+                                                       indices_data,
+                                                       sizeD,
+                                                       isizeH, isizeW,
+                                                       osizeH, osizeW);
+      }
+    );
+  }
+  else
+  {
+    int64_t b;
+#pragma omp parallel for private(b)
+    for (b = 0; b < sizeB; b++)
+    {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d_backward", [&] {
+        /* get raw pointers */
+        scalar_t *gradInput_data = gradInput.data<scalar_t>();
+        scalar_t *gradOutput_data = gradOutput.data<scalar_t>();
+        int64_t *indices_data = indices.data<int64_t>();
+
+        adaptive_max_pool2d_backward_out_frame<scalar_t>(gradInput_data+b*sizeD*isizeH*isizeW,
+                                                         gradOutput_data+b*sizeD*osizeH*osizeW,
+                                                         indices_data+b*sizeD*osizeH*osizeW,
+                                                         sizeD,
+                                                         isizeH, isizeW,
+                                                         osizeH, osizeW);
+        }
+      );
+    }
+  }
+
+  return gradInput;
+}
+
+} // namespace
+
+std::tuple<Tensor&, Tensor&> adaptive_max_pool2d_out_cpu(
+  Tensor& output,
+  Tensor& indices,
+  const Tensor& input,
+  IntArrayRef output_size)
+{
+  adaptive_max_pool2d_out_cpu_template(
+    output,
+    indices,
+    input,
+    output_size);
+  return std::tuple<Tensor&, Tensor&>(output, indices);
+}
+
+std::tuple<Tensor, Tensor> adaptive_max_pool2d_cpu(
+  const Tensor& input,
+  IntArrayRef output_size)
+{
+  Tensor output = at::empty({0}, input.options());
```

Comment:
I think it also could use `auto` here, such as https://github.com/pytorch/pytorch/pull/14714/files#diff-257f62854d62776a19a5480585a39962R299 not sure what would be the best practice here.

Reply:
I don't know the style guide,  this one is from `FractionalMaxPool2d.cpp`.

Reply:
I don't really mind, but I know @mike7ant prefers it when you don't use auto ;)